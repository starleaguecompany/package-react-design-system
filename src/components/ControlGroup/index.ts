import ControlGroup from './src/ControlGroup'
import { ControlGroupProps } from './types/ControlGroup.types'

export { ControlGroup, ControlGroupProps }
