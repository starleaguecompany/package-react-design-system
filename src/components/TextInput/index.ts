import TextInput from './src/TextInput';
import { TextInputProps } from './types/TextInput.types';

export { TextInput, TextInputProps };
