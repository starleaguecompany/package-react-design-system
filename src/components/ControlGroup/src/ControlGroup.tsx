import * as React from 'react';
import { useStyles, error } from '@starleaguecompany/package-react-utils';

import { Space } from '../../Space';

import { ControlGroupProps } from '../types/ControlGroup.types';
import styles from '../styles/ControlGroup.module.less';

/**
 * @description ControlGroup component.
 *
 * @component
 * @example
 * ```jsx
 * <ControlGroup>
 *  <TextInput placeholder="Left input position"/>
 *  <TextInput placeholder="Center input position"/>
 *  <TextInput placeholder="Right input position"/>
 * </ControlGroup>
 * ```
 */
const ControlGroup = React.forwardRef<HTMLDivElement, ControlGroupProps>((props, ref) => {
  const { className, children, direction, ...restProps } = props;

  const cx = useStyles(styles);
  const count = React.useMemo(() => React.Children.count(children), [children]);
  const condition = count < 2;

  // Ensure group has more than one child node
  if (condition) {
    error({
      condition,
      message: 'Ensure group has more than one child node',
    })();
  }

  const group = React.useMemo(() => {
    return React.Children.map(children as React.ReactElement, (child, index) => {
      if (condition) {
        return child;
      }

      let positionInGroup;

      if (index === 0) {
        positionInGroup = 'start';
      } else if (index === count - 1) {
        positionInGroup = 'end';
      } else {
        positionInGroup = 'middle';
      }

      return React.cloneElement(child, { ...child.props, positionInGroup, directionGroup: direction });
    });
  }, [children]);

  return (
    <div ref={ref} data-qa="ControlGroup" className={cx(className, 'container')} {...restProps}>
      <Space direction={direction} size={2}>
        {group}
      </Space>
    </div>
  );
});

ControlGroup.defaultProps = {
  direction: 'horizontal',
};

export default ControlGroup;
