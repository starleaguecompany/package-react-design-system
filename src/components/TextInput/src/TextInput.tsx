import * as React from 'react';
import {
  useStyles,
  useBoolean,
  useMergedRef,
  useFormControlContext,
  safeInvoke,
  isUndefined,
} from '@starleaguecompany/package-react-utils';

import { Icon } from '../../Icon';
import { Spinner } from '../../Spinner';

import { TextInputProps } from '../types/TextInput.types';
import styles from '../styles/TextInput.module.less';

export function fixControlledValue<T>(value: T): T | '' {
  if (typeof value === 'undefined' || value === null) {
    return '';
  }

  return value;
}

/**
 * @description The Text Input component allows user to type in text
 *
 * @component
 * @example
 * ```jsx
 * <TextInput name="field" value="value" />
 * ```
 */
const TextInput = React.forwardRef<HTMLInputElement, TextInputProps>((props, ref) => {
  const {
    autoFocus,
    defaultValue,
    value,
    label,
    placeholder,
    postfix,
    icon,
    disabled,
    invalid,
    readOnly,
    loading,
    positionInGroup,
    directionGroup,
    onFocus,
    onBlur,
    onChange,
    className,
    type,
    ...restProps
  } = props;

  const cx = useStyles(styles);
  const formControlContext = useFormControlContext();
  const inputRef = React.useRef<HTMLInputElement>();
  const [focused, setFocused] = useBoolean(autoFocus || false);
  const [filled, setFilled] = useBoolean(value !== undefined && value !== '');

  const isActive = focused && !readOnly;
  const isDisabled = formControlContext.disabled || disabled;
  const isInvalid = formControlContext.invalid || invalid;
  const hasIcon = icon || loading;

  const classNames = cx(className, 'container', {
    focused,
    active: isActive,
    disabled: isDisabled,
    invalid: isInvalid,
    'with-icon': hasIcon,
    [`group-${directionGroup}`]: directionGroup,
    [`group-${positionInGroup}`]: positionInGroup,
  });

  const classNamesLabel = cx('placeholder', {
    active: isActive,
    filled: filled,
    disabled: isDisabled,
    invalid: isInvalid,
    'with-icon': hasIcon,
  });

  const classNamesInput = cx('input', {
    active: isActive,
    filled: filled,
    disabled: isDisabled,
    invalid: isInvalid,
  });

  const classNamesInputSegment = cx('inputSegment', {
    centered: isUndefined(label) && (filled || focused),
  });

  const iconElement = React.useMemo(() => {
    if (loading) {
      return <Spinner className={cx('icon')} />;
    }

    if (icon) {
      return <Icon className={cx('icon')} icon={icon} />;
    }

    return null;
  }, [icon, loading]);

  const postfixElement = React.useMemo(() => {
    if (postfix && filled) {
      return <div className={cx('inputPostfix')}>&nbsp;{postfix}</div>;
    }

    return null;
  }, [postfix, filled]);

  const labelElement = React.useMemo(() => {
    if (label) {
      return <label className={classNamesLabel}>{label}</label>;
    }

    return null;
  }, [label, filled, focused]);

  const placeholderElement = React.useMemo(() => {
    if (label || filled || focused) {
      return null;
    }

    if (placeholder) {
      return <label className={classNamesLabel}>{placeholder}</label>;
    }

    return null;
  }, [label, placeholder, filled, focused, isDisabled]);

  const handleClick = React.useCallback(
    (_event: React.MouseEvent<HTMLDivElement>) => {
      inputRef.current?.focus();
      setFocused.on();
    },
    [inputRef]
  );

  const handleFocus = React.useCallback(
    (event: React.FocusEvent<HTMLInputElement>) => {
      setFocused.on();
      safeInvoke(onFocus, event);
    },
    [onFocus]
  );

  const handleBlur = React.useCallback(
    (event: React.FocusEvent<HTMLInputElement>) => {
      setFocused.off();
      safeInvoke(onBlur, event);
    },
    [onBlur]
  );

  const handleChange = React.useCallback(
    event => {
      safeInvoke(onChange, event);
    },
    [onChange]
  );

  React.useEffect(() => {
    if (value !== undefined && value !== '') {
      setFilled.on();
    } else {
      setFilled.off();
    }
  }, [value]);

  return (
    <div data-qa="TextInput" className={classNames} onClick={handleClick}>
      <div className={classNamesInputSegment}>
        <div className={cx('inputWrapper')}>
          <div className={cx('inputValue')}>{fixControlledValue(value)}</div>
          <input
            autoFocus={autoFocus}
            spellCheck="false"
            autoComplete="off"
            ref={useMergedRef(inputRef, ref)}
            type={type}
            value={fixControlledValue(value)}
            disabled={isDisabled}
            readOnly={readOnly}
            className={classNamesInput}
            onFocus={handleFocus}
            onBlur={handleBlur}
            onChange={handleChange}
            {...restProps}
          />
          {postfixElement}
        </div>
      </div>
      {labelElement}
      {placeholderElement}
      {iconElement}
    </div>
  );
});

TextInput.defaultProps = {
  type: 'text',
};

export default TextInput;
