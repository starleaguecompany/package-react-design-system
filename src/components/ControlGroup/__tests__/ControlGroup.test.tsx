import * as React from 'react'
import { render, screen } from '@testing-library/react'

import { TextInput } from '../../TextInput'
import { ControlGroup } from '../../ControlGroup'

describe('ControlGroup', () => {
  test('should render correctly with no props', () => {
    render(<ControlGroup />)

    expect(screen.getByTestId('ControlGroup')).toMatchSnapshot()
    expect(screen.getByTestId('ControlGroup')).toHaveClass('container', { exact: true })
  })

  test('should render correctly with attributes', () => {
    render(<ControlGroup id="test-id" className="test-class" />)

    expect(screen.getByTestId('ControlGroup')).toMatchSnapshot()
    expect(screen.getByTestId('ControlGroup')).toHaveAttribute('id', 'test-id')
    expect(screen.getByTestId('ControlGroup')).toHaveClass('test-class')
  })

  test('should render correctly with direction prop', () => {
    render(
      <ControlGroup direction="vertical">
        <TextInput placeholder="Название поля 1" />
        <TextInput placeholder="Название поля 2" />
        <TextInput placeholder="Название поля 3" />
      </ControlGroup>
    )

    expect(screen.getByTestId('ControlGroup')).toMatchSnapshot()
    expect(screen.getAllByTestId('TextInput')[0]).toHaveClass('group-vertical group-start')
    expect(screen.getAllByTestId('TextInput')[1]).toHaveClass('group-vertical group-middle')
    expect(screen.getAllByTestId('TextInput')[2]).toHaveClass('group-vertical group-end')
  })

  test('should render correctly with children', () => {
    render(
      <ControlGroup>
        <TextInput placeholder="Название поля 1" />
        <TextInput placeholder="Название поля 2" />
        <TextInput placeholder="Название поля 3" />
      </ControlGroup>
    )

    expect(screen.getByTestId('ControlGroup')).toMatchSnapshot()
    expect(screen.getByTestId('ControlGroup')).toContainElement(screen.getAllByTestId('TextInput')[0])
    expect(screen.getByTestId('ControlGroup')).toContainElement(screen.getAllByTestId('TextInput')[1])
    expect(screen.getByTestId('ControlGroup')).toContainElement(screen.getAllByTestId('TextInput')[2])
    expect(screen.getAllByTestId('TextInput')[0]).toHaveClass('group-horizontal group-start')
    expect(screen.getAllByTestId('TextInput')[1]).toHaveClass('group-horizontal group-middle')
    expect(screen.getAllByTestId('TextInput')[2]).toHaveClass('group-horizontal group-end')
  })
})
