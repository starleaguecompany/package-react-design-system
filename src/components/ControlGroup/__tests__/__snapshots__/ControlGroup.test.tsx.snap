// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlGroup should render correctly with attributes 1`] = `
<div
  class="test-class container"
  data-qa="ControlGroup"
  id="test-id"
>
  <div
    class="container direction-horizontal"
    data-qa="Space"
    style="column-gap: 2px; row-gap: 2px;"
  />
</div>
`;

exports[`ControlGroup should render correctly with children 1`] = `
<div
  class="container"
  data-qa="ControlGroup"
>
  <div
    class="container direction-horizontal"
    data-qa="Space"
    style="column-gap: 2px; row-gap: 2px;"
  >
    <div
      class="container group-horizontal group-start"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
      <label
        class="placeholder"
      >
        Название поля 1
      </label>
    </div>
    <div
      class="container group-horizontal group-middle"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
      <label
        class="placeholder"
      >
        Название поля 2
      </label>
    </div>
    <div
      class="container group-horizontal group-end"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
      <label
        class="placeholder"
      >
        Название поля 3
      </label>
    </div>
  </div>
</div>
`;

exports[`ControlGroup should render correctly with direction prop 1`] = `
<div
  class="container"
  data-qa="ControlGroup"
>
  <div
    class="container direction-vertical"
    data-qa="Space"
    style="column-gap: 2px; row-gap: 2px;"
  >
    <div
      class="container group-vertical group-start"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
      <label
        class="placeholder"
      >
        Название поля 1
      </label>
    </div>
    <div
      class="container group-vertical group-middle"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
      <label
        class="placeholder"
      >
        Название поля 2
      </label>
    </div>
    <div
      class="container group-vertical group-end"
      data-qa="TextInput"
    >
      <div
        class="inputSegment"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="inputValue"
          />
          <input
            autocomplete="off"
            class="input"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
      <label
        class="placeholder"
      >
        Название поля 3
      </label>
    </div>
  </div>
</div>
`;

exports[`ControlGroup should render correctly with no props 1`] = `
<div
  class="container"
  data-qa="ControlGroup"
>
  <div
    class="container direction-horizontal"
    data-qa="Space"
    style="column-gap: 2px; row-gap: 2px;"
  />
</div>
`;
